File: backend/python/authorize_purchases.py
from flask import Flask, request, jsonify

app = Flask(__name__)

# Securely store API keys outside of this file.
# For example purposes only, a sample API key is
# used below.

API_KEY = "cad1"

@app.route('/purchase/<player_id>', methods=['POST'])
def handle_transaction(player_id):
    # Validate the API key before processing the call
    api_key = request.headers.get('x-api-key')
    if not api_key or api_key != API_KEY:
        return jsonify({"error": "Unauthorized"}), 401
    if request.is_json:
        try:
            # Parse the JSON content from the request
            data = request.get_json()
            # Access transactionId and items from the JSON object
            transaction_id = data['transactionId']
            items = data['items']
            # This is where you'd trigger logic for granting
            # the player their items.
            # This example just prints the player ID, transaction ID,
            # and the items purchased.
            print(f'Player ID: {player_id}')
            print(f'Transaction ID: {transaction_id}')
            for item in items:
                print(f'Item ID: {item["id"]}, Quantity: {item["quantity"]}')
            # Respond with a 200 OK status after you grant the player their
            # items. This notifies Stash that we can charge the player.
            return jsonify({"message": "Transaction processed successfully"}), 200
        except KeyError as e:
            # If required keys are missing, respond with a 400 Bad Request
            return jsonify({"error": f"Missing key: {str(e)}"}), 400
    else:
        # If the request content-type is not JSON, respond with a 400 Bad Request
        return jsonify({"error": "Request must be JSON"}), 400

if __name__ == '__main__':
    app.run(debug=True)

File: backend/javascript/authorize_purchases.js
const express = require('express');
const app = express();
const PORT = 3000;

// Middleware to parse JSON bodies.
app.use(express.json());

// Validate the API key before processing the call.
// Securely store API keys outside of this file.
// For example purposes only, a sample API key is
// used below.
const apiKeyValidation = (req, res, next) => {
    const apiKey = req.header('x-api-key');
    if (!apiKey || apiKey !== 'abc123') {
        return res.status(401).json({ error: 'Invalid API key' });
    }
    next();
};

// Endpoint to handle purchase requests.
app.post('/purchase/:playerId', apiKeyValidation, (req, res) => {
    const playerId = req.params.playerId;
    const transactionData = req.body;

    // Basic error handling for missing or invalid JSON.
    if (!transactionData || !transactionData.transactionId || !transactionData.items) {
        return res.status(400).json({ error: 'Invalid or missing JSON data' });
    }

    // This is where you'd trigger logic for granting
    // the player their items.
    // This example prints the player ID and transaction data.

    console.log(`Player ID: ${playerId}`);
    console.log('Transaction Data:', transactionData);

    res.status(200).json({ message: 'Purchase processed successfully' });
});

// Error handling middleware for invalid JSON.
app.use((err, req, res, next) => {
    if (err instanceof SyntaxError && err.status === 400 && 'body' in err) {
        return res.status(400).json({ error: 'Invalid JSON' });
    }
    next();
});

// Start the server
app.listen(PORT, () => {
    console.log(`Server is running on port ${PORT}`);
});